Compile full DELETE with JOIN including ORDER BY and LIMIT in MySQL grammar

tegos opened the pull request:
### Problem

By default, Laravel silently ignores `ORDER BY` and `LIMIT` on `DELETE … JOIN` queries when using the MySQL grammar.
This results in unexpected behavior: instead of deleting rows in small batches, the query deletes *all* matching rows at once.

Example:

```php
$batchSize = 500;
$totalDeleted = 0;

do {
	$deleted = DB::table('cross_product_references as cpr')
		->leftJoin('products as p1', fn($join) => $join
			->on('cpr.article', '=', 'p1.article')
			->on('cpr.brand_id', '=', 'p1.brand_id'))
		->leftJoin('products as p2', fn($join) => $join
			->on('cpr.cross_article', '=', 'p2.article')
			->on('cpr.cross_brand_id', '=', 'p2.brand_id'))
		->where(fn($query) => $query
			->whereNull('p1.id')
			->orWhereNull('p2.id'))
		->limit($batchSize)
		->orderBy('cpr.id')
		->delete();
	$totalDeleted += $deleted;
} while ($deleted > 0);

$this->line("Deleted <info>$totalDeleted</info> rows.");
```

**Expected:** delete in batches of 500
**Actual:** deletes all 500k+ rows in one query

---

### Fix

* Added `compileDeleteWithJoins` override in `MySqlGrammar`.
* Now compiles full SQL with `JOIN`, `ORDER BY`, and `LIMIT` instead of stripping them.
* Leaves it to the underlying MySQL/MariaDB/Vitess implementation to decide whether the query is valid.

---

### Why

* Avoids silent ignoring of `ORDER BY` and `LIMIT`, which could lead to **accidental mass deletions**.
* Provides consistent behavior: Laravel now generates the full query requested by the developer.
* Compatible with systems like Vitess, which do support `DELETE … JOIN … ORDER BY/LIMIT`.
* On vanilla MySQL, developers will now get `QueryException`.

taylorotwell committed:
Update CHANGELOG

timacdonald committed:
Ensure cached and uncached routes share same precedence when resolving actions and names (#56920)

* Ensure cached and uncached routes share same precedence

* formatting

* Formatting

* formatting

* Fix tests

AhmedAlaa4611 committed:
Re-enable previously commented assertions (#56930)

AhmedAlaa4611 committed:
Reorder .gitignore entries for consistency and readability (#56963)

stancl committed:
[12.x] SQLite: Allow setting any pragmas (#56962)

* SQLite: Allow setting any pragmas

* formatting

---------

Co-authored-by: Taylor Otwell <taylor@laravel.com>

alipowerful7 committed:
refactor: remove unused array from docblock (#56961)

moshe-autoleadstar committed:
PendingResourceRegistration withoutMiddleware never returns array (#56959)

SjorsO committed:
allow fakerphp/faker to not be installed (#56953)

AhmedAlaa4611 committed:
Fix Validator placeholderHash PHPDoc (#56947)

Muffinman committed:
Handle MariaDB innodb_snapshot_isolation=ON (#56945)

Fixes #56944

tuandp committed:
[12.x] Add PhpRedis pack ignore numbers option (#56941)

* Add phpredis pack ignore numbers support

* Update PhpRedisConnector.php

---------

Co-authored-by: Taylor Otwell <taylor@laravel.com>

realpvz committed:
test(support): add edge-case tuples for preg_replace_array (#56937)

gehrisandro committed:
[12.x] Allow for BackedEnum on dynamic blade component (#56940)

* Allow for BackedEnum on dynamic blade component

* use enum_value() helper function

Co-authored-by: Andrew Brown <browner12@gmail.com>

* Fix missing function import

---------

Co-authored-by: Andrew Brown <browner12@gmail.com>

vincentvanhoven committed:
[12.x] Remove one redundant array access (#56931)

* [REMOVE] Redundant array access

* [retrigger tests]

* Revert "[retrigger tests]"

This reverts commit 9e83eb8b704aa7a46d9b72e8e56eb1453ef939da.

theHocineSaad committed:
[12.x] Add withoutGlobalScopesExcept() to keep only specified global scopes (#56957)

* Add method to remove all except specified global scopes

* Fix parameter type hint

* Add test for removing all global scopes except specified ones

* Update Builder.php

---------

Co-authored-by: Taylor Otwell <taylor@laravel.com>

AhmedAlaa4611 committed:
Make visibility consistent (#56970)

shaedrich committed:
Change list to tuple in PHPDoc block (#56967)

cosmastech committed:
aggregate service provider docblocks (#56968)

thojo0 committed:
[12.x] add --whisper option to schedule:work command (#56969)

* add --wisper option to schedule:work

* Update ScheduleWorkCommand.php

---------

Co-authored-by: Taylor Otwell <taylor@laravel.com>

AhmedAlaa4611 committed:
Update Faker suggestion to match skeleton version (#56974)

browner12 committed:
remove unnecessary `with()` helper call (#56975)

if you're not passing a second parameter with a closure to `with()`, it doesn't do anything other than return the value. this is verrrry old legacy code from 10+ years ago that isn't necessary anymore to chain.

AhmedAlaa4611 committed:
Config: Move some items into pragmas (#56980)

kamilkozak committed:
Add callback support to takeUntilTimeout in LazyCollection (#56981)

* Add callback support to takeUntilTimeout in LazyCollection

The takeUntilTimeout method now accepts an optional callback that is invoked when the timeout is reached, providing the last yielded value and key. Tests and type assertions have been updated to cover the new callback functionality.

* Fix indentation

* Simplify initialization of `$timedOutWith` in SupportLazyCollectionTest

AhmedAlaa4611 committed:
Utilize is_finite() (#56990)

cosmastech committed:
property promotion (#56989)

browner12 committed:
do not use `with()` helper when no second argument is passed (#56986)

if no second argument is passed to `with()` it simply returns the value, so there's no reason to call it when we only pass 1 argument.

while the performance impact it probably negligible, it does take 1 thing out of the call stack, which is helpful.

sethsandaru committed:
fixed wrong type (#56987)

theHocineSaad committed:
[12.x] Some quick fixes (#56991)

* fix: correct typo in docblock for setName method

* fix: update return type in getRouteDomain method to string|null

realpvz committed:
tests: Ensure transaction callbacks run in FIFO order (#56973)

fritz-c committed:
pass attributes and parent arguments to Factory Sequence as in ->state() (#56972)

jrseliga committed:
support `Castable` on `Enum` (#56977)

browner12 committed:
add trailling commas in multline method signatures (#56992)

helps with better diffs

used the following regex to find them, not perfect, but it gets you close:

```
[^,]
(\ *)\)\ \{
```

amirhshokri committed:
Improve docblocks for nullable parameters (#56995)

amirhshokri committed:
Improve docblocks for nullable parameters (#56996)

amirhshokri committed:
Improve docblocks for nullable parameters (#56997)

taylorotwell committed:
Revert "Config: Move some items into pragmas (#56980)" (#57003)

This reverts commit 52a1f6e07f88f94ff9019c4836cc794d95257afa.

joaopalopes24 committed:
[12.x]: Cache Session Driver (#56887)

* feat(session): adding new method to remeber value for a given time

* wip

* feat(cache): adding session driver skeleton

* feat(cache): finish cache session driver

* test(cache): adding tests for cache session driver

* wip

* wip

* wip

* wip

* refactor(cache): some adjusts

* wip

* formatting

* refactor(cache): some adjusts in cache session driver

* Update cache.php

* Update CacheManager.php

* formatting

---------

Co-authored-by: Taylor Otwell <taylor@laravel.com>

taylorotwell committed:
Update facade docblocks

Lukasss93 committed:
[12.x] Add support for #[UseResource(...)] and #[UseResourceCollection(...)] attributes on models (#56966)

* test: add fixture resources for Eloquent resource collection tests

* feat: add UseResource and UseResourceCollection attributes for model resource transformation

* refactor: update parameter type hints for resource and resource collection attributes

* formatting

---------

Co-authored-by: Taylor Otwell <taylor@laravel.com>

crynobone committed:
[12.x] Test Improvements (#57010)

Signed-off-by: Mior Muhammad Zaki <crynobone@gmail.com>

crynobone committed:
[12.x] Test Improvements (#57031)

* [12.x] CI Improvements

Signed-off-by: Mior Muhammad Zaki <crynobone@gmail.com>

* wip

Signed-off-by: Mior Muhammad Zaki <crynobone@gmail.com>

---------

Signed-off-by: Mior Muhammad Zaki <crynobone@gmail.com>

daniser committed:
[12.x] Infinite method chaining in contextual binding builder (#57026)

Co-authored-by: Sergey Danilchenko <s.danilchenko@ttbooking.ru>

daniser committed:
[12.x] Improved manager typehints (#57024)

Co-authored-by: Sergey Danilchenko <s.danilchenko@ttbooking.ru>

dependabot[bot] committed:
Bump vite in /src/Illuminate/Foundation/resources/exceptions/renderer (#57009)

Bumps [vite](https://github.com/vitejs/vite/tree/HEAD/packages/vite) from 5.4.19 to 5.4.20.
- [Release notes](https://github.com/vitejs/vite/releases)
- [Changelog](https://github.com/vitejs/vite/blob/v5.4.20/packages/vite/CHANGELOG.md)
- [Commits](https://github.com/vitejs/vite/commits/v5.4.20/packages/vite)

---
updated-dependencies:
- dependency-name: vite
  dependency-version: 5.4.20
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

taylorotwell committed:
Update facade docblocks

xurshudyan committed:
Correct APC cache store docblock types (#57020)

Co-authored-by: Xurshudyan <v.khurshudyan@aico.swiss>

taylorotwell committed:
fix type hint

glioympas committed:
Allow dynamic tries() method on Queueable listeners (#57014)

dxnter committed:
[12.x] Add --json option to ScheduleListCommand (#57006)

* feat: add `--json` option to `schedule:list`

* formatting

---------

Co-authored-by: Taylor Otwell <taylor@laravel.com>

StyleCIBot committed:
Apply fixes from StyleCI

browner12 committed:
[12.x] `with()` helper call simplification (#57041)

* remove unecessary `with()` helper calls

when we have a well defined closure, there is no reason to use the `with()` helper, or even the closure at all. in most cases, but using a temporary variable we achieve the same effect with arguably cleaner easier to read code.

* style fixes

* revert refactor

I can't figure out how to fix the tests right now, so reverting these

browner12 committed:
handle all Enum types for default values (#57040)

by using the internal `enum_value()` helper, we can handle both `UnitEnum`s and `BackedEnum`s, similar to how we do elsewhere in the framework.

we check for `UnitEnum`, because it is the broader of the 2 types.

AhmedAlaa4611 committed:
Refactor chained method calls for readability (#57050)

AhmedAlaa4611 committed:
Improve docblock wording (#57056)

AhmedAlaa4611 committed:
Refactor chained method calls for readability (#57054)

avosalmon committed:
[12.x] Update local exception page (#57036)

* Configure Vite with Tailwind CSS

* Register new renderer template

* Base layout with static content

* Use dashed border

* Add separator

* Replace hard-coded values

* Replace hard-coded query

* Add empty state

* Clean up section container

* Clean up

* Add section-container

* Use section-container in topbar

* Clean up

* Extract query component

* Add request-header component

* Add request-body component

* Add routing component

* Add routing-parameter component

* Add overview component

* Add request-url component

* Add header component

* Add topbar component

* Add trace component skeleton

* Group exception frames

* Add formatted-source and file-with-line components

* Add frames

* Add previous frame

* Show code snippet without syntax highlight

* Use exceptionAsMarkdown

* Remove unused props

* Assign snippet to a variable

* Syntax highlight code snippet

* Update phiki

* Syntax highlight query, body, and route parameters

* Update phiki

* Use LineDecoration to highlight code snippet

* Unescape highlighted text

* Add syntax-highlight component

* Handle files with less than 5 lines

* Install Alpine.js

* Add layout component

* Add laravel-ascii component

* Update method badge

* Mark frame as main

* Expand/collapse frames

* Use current color in SVGs

* Don't show callable when the frame doesn't have a class

* Copy request URL to clipboard

* Truncate long source

* Add tooltip

* wip: tooltip

* Add side prop to tooltip

* Add tooltip on request url

* Prevent non-vendor frame from overflowing

* Allow horizontal scroll on request body

* Add tooltip on request headers

* Add tooltip on database query

* Hide JS-dependent elements until Alpine is loaded

* Paginate queries

* Add database icon

* Add folder icons

* Add copy icon

* Add globe icon

* Add alert icon

* Copy exception as markdown

* Add badge component

* Add http-method component

* Expand frames on clicking parent div

* Open file in editor

* Use pointer cursor on pagination buttons

* Replace custom tooltip component with tippy.js

* Upgrade phiki/phiki

* Use dark-plus theme

* Use dvh instead of screen

* wip: light mode

* Style gutter text

* Only show the first 100 queries

* light mode styling

* Add light theme for syntax highlighter

* Uppercase keys

* Style copy as markdown button

* Clean up

* Add hover effect on laravel ascii logo

* Add shadow

* Fix tooltip position for queries

* Tweak padding for mobile

* Allow HTML in tooltip

* Show frame arguments

* Syntax highlight frame source

* Add markdown template

* Replace renderer directory

* Simplify CSS rendering

* Remove unused method

* Revert unintended changes

* Rename array key

* Rename type to operator

* Set max width for tooltip

* Display vendor frames in two lines

* Apply fixes from StyleCI

* Add empty state for routing context

* Only round top corners when frame is expanded

* Update dot color for non-vendor frame

* Update topbar height

* Adjust padding around header section

* Move up request url and have it overlap the separator line

* Adjust spacing

* Adjust spacing for mobile

* Replace shadow-sm with shadow-xs

* Apply bg-white without opacity in light mode

* Adjust spacing

---------

Co-authored-by: StyleCI Bot <bot@styleci.io>

jackbayliss committed:
[12.x] Add ability to control QueueWorker memory exceeded exit code (#57044)

* init

* Update WorkCommand.php

* cs

* fix up test

* init

* Update WorkCommandTest.php

* try fix flakiness

* cs

* Update WorkCommandTest.php

* Update WorkCommandTest.php

* readd const to prevent breaking change

:trollface:

* use variable

---------

Co-authored-by: Taylor Otwell <taylor@laravel.com>

PeteBishwhip committed:
[12.x] Ensure `laravel-cloud-socket` respects `LOG_LEVEL` (#57071)

* Add log level to laravel-cloud-socket configuration

* Add test to ensure log levels are respected

---------

Co-authored-by: Pete Bishop <peter@soniaandpete.com>

taylorotwell committed:
Update CHANGELOG

hebbet committed:
[12.x] Allow new versions for phiki/phiki than 2.0.0 (#57075)

jackbayliss committed:
[12.x] Use null coalescing for memoryExceededExitCode (#57090)

* Update Worker.php

* Update Worker.php

* Update Worker.php

* use 0

pdewit committed:
Fix 'can' function defined in RouterRegistrar in #54648 but did not work (#57072)

patrickcarlohickman committed:
[12.x] Fix SQS FIFO and fair queue support (#57080)

* Add tests to ensure that job objects are properly queued on SQS using push or pending dispatch.

* Add tests to ensure that job objects can properly enable SQS fair queues.

* Add a test to ensure that job strings are properly pushed onto SQS FIFO queues.

* Add tests to ensure that job objects are properly pushed onto SQS FIFO queues using push or pending dispatch.

* Add tests to ensure that DelaySeconds are not sent to SQS FIFO queues when delayed.

* Properly handle the message group and deduplication queable options for SQS FIFO and fair queues.

* Ensure DelaySeconds is not sent to SQS FIFO queues.

* Restore normal uuid creation at end of tests.

* formatting

---------

Co-authored-by: Taylor Otwell <taylor@laravel.com>

taylorotwell committed:
atomically flush redis cache tags (#57098)

shaedrich committed:
Add type hints to \Illuminate\Support\Str (#57096)

glensc committed:
Update Database Connection getElapsedTime comment to specify unit (#57099)

To understand whether the value is seconds, milliseconds, microseconds, nanoseconds

BinaryKitten committed:
[12.x] Add support for Ordinal Position in validation messages (#57109)

* Add Support for Ordinal Position in validation messages

* update commented out item to match test.

* Remove commented code from new test - its not been used from the original since v9 no point adding it here

* Empty commit to re-trigger failed CI

avosalmon committed:
[12.x] Fix exception frame file path on Windows (#57103)

* Replace forward slash with DIRECTORY_SEPARATOR

* Use join_paths helper

* Add unit tests for Frame class

joaokamun committed:
Add fallback to copy buttons on new exception page (#57092)

* Add fallback to copy

* Update layout.blade.php

* Move global function to scripts.js

* Format

---------

Co-authored-by: Taylor Otwell <taylor@laravel.com>
Co-authored-by: Ryuta Hamasaki <avosalmon@gmail.com>

1tim22 committed:
Adds `Macroable` trait to `Illuminate/Support/Benchmark` (#57107)

Updates `SupportBenchmarkTest` to test `Benchmark` macro registration

Arbitrary macros may now decorate `Benchmark` to customize how
benchmarks are reported or logged.

Co-authored-by: Andrew Baron <abaron@sbts.edu>

taylorotwell committed:
Update CHANGELOG

BinaryKitten committed:
Apply intl extension check to ordinal position (#57112)

taylorotwell committed:
Update CHANGELOG

crynobone committed:
Update `package-lock.json`

dependabot[bot] committed:
Bump vite in /src/Illuminate/Foundation/resources/exceptions/renderer (#57114)

Bumps [vite](https://github.com/vitejs/vite/tree/HEAD/packages/vite) from 7.1.2 to 7.1.6.
- [Release notes](https://github.com/vitejs/vite/releases)
- [Changelog](https://github.com/vitejs/vite/blob/main/packages/vite/CHANGELOG.md)
- [Commits](https://github.com/vitejs/vite/commits/v7.1.6/packages/vite)

---
updated-dependencies:
- dependency-name: vite
  dependency-version: 7.1.6
  dependency-type: direct:development
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

PerryvanderMeer committed:
short hand false props (#57104)

platoindebugmode committed:
Allow Number parsing helpers to return false (#57127)

cosmastech committed:
[12.x] Refactor `RedisTaggedCache@flush()` to allow for custom connections (#57122)

* Refactor flush method to improve redisPrefix retrieval

Refactor redisPrefix assignment to use instance checks.

* clean up

StyleCIBot committed:
Apply fixes from StyleCI

pxlrbt committed:
Use light-dark scheme for exception renderer (#57128)

abdelrahmenAyman committed:
fix: replace log function and logger helper return type ?LogManager with ?LoggerInterface (#57028)

tegos committed:
Add MySQL-specific constraints for delete statements with joins

tegos committed:
Add tests to validate MySQL behavior for DELETE with JOIN restrictions

barclaymichael committed:
Fix session value is missing assertion (#57134)

artumi-richard committed:
median() div swapped for intdiv() (#57148)

Finding the middle of a list of values using the float div operator (/) is wrong, and goes wrong at some higher values where the int/2 is not representable as a float to single digit precision. 

That's why intdiv() exists.

IonBazan committed:
[12.x] Fix PHP 8.5 null-key deprecations (#57137)

* Fix PHP 8.5 null-key deprecations

* Fix getOrPut

* Update InteractsWithIO.php

* Update Collection.php

---------

Co-authored-by: Taylor Otwell <taylor@laravel.com>

taylorotwell committed:
Update CHANGELOG

taylorotwell committed:
Revert "short hand false props (#57104)" (#57151)

This reverts commit af710f5d8da37173cb8fb0310727691a50b71756.

taylorotwell committed:
Update CHANGELOG

tegos committed:
chore: update tests for DELETE statements with and without JOIN using LIMIT

tegos committed:
chore: exclude mariadb from LIMIT on DELETE with JOIN test

GrahamCampbell commented:
Why not just generate the invalid statement always, so that you get the error back from MySQL?

tegos commented:
@GrahamCampbell
True, Vitess/PlanetScale support this by rewriting the query internally,
but that's not native MySQL behavior. 😄 
We're targeting the standard MySQL driver here, not clustering systems or proxies that add their own SQL semantics.
With the current `MySqlGrammar`, Laravel would silently drop `LIMIT`/`ORDER BY` on deletes with joins, potentially removing all rows.
By throwing an explicit exception, developers can immediately see the problem and apply the known workaround
(e.g. select IDs with `FOR UPDATE` then delete with `WHERE IN`). This keeps Laravel's behavior consistent and predictable.

GrahamCampbell commented:
I still think it's better to just allow the server to complain the query is invalid.

cosmastech committed:
[12.x] fix static analysis error  (#57162)

* fix static analysis error

* fix typehint instead

Carnicero90 committed:
Fix: Handle non-string returns from Htmlable::toHtml() in e() helper (#57157)

This adds null coalescing to maintain backward compatibility when
Htmlable::toHtml() returns non-string values.

**Breaking Change Context:**
The addition of explicit `: string` return type creates a BC break because
Htmlable::toHtml() only has PHPDoc type hints, not runtime enforcement.
 IMHO it should probably be reverted, this commit provides at least some backward compatibility.

**Example of previously valid code that now breaks:**
```php
class CustomHtmlable implements Htmlable
{
    public function toHtml()
    {
        // Valid before: PHPDoc says @return string but no runtime enforcement
        return $this->hasContent() ? $this->content : null;
    }
}

// This worked before but now causes TypeError
echo e(new CustomHtmlable());

jamessa committed:
[12.x] Fix pending attributes in schedule group (#57156)

* use copy instead of reference

* style fix

barryvdh committed:
Remove Request overview from Exceptions (#57158)

* Remove Request overview from Exceptions

* Fix HTTP status code

taylorotwell committed:
convert encoding

daniser committed:
[12.x] Pass "throw" option from scoped to parent disk (#57163)

Co-authored-by: Sergey Danilchenko <s.danilchenko@ttbooking.ru>

parijke committed:
Fix: Update return type annotation for array() method to always return an array (#57164)

taylorotwell committed:
fix grouping

daniser committed:
[12.x] Adjust `Arr` typehints (#57165)

Co-authored-by: Sergey Danilchenko <s.danilchenko@ttbooking.ru>

taylorotwell commented:
I sort of agree that maybe we should just add the LIMIT and if the underlying MySQL system doesn't like it then it will error anyways? But, those using PlanetScale would still be able to use that feature?

tegos committed:
feat: support ORDER BY and LIMIT in MySQL DELETE without joins, update joined clause handling

tegos committed:
test: update MySQL DELETE with JOIN tests to support ORDER BY and LIMIT

taylorotwell committed:
Update MySqlGrammar.php

tegos commented:
I'll close this PR and open a new one against master with a clean branch based on the latest upstream.
